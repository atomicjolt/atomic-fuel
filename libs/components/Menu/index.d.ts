declare function Menu(props: any): any;
declare namespace Menu {
    namespace propTypes {
        const renderChildren: PropTypes.Validator<(...args: any[]) => any>;
    }
}
export default Menu;
import PropTypes from "prop-types";
