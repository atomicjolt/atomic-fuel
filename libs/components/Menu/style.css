@keyframes spinner {
  0% {
    transform: rotate(0deg); }
  100% {
    transform: rotate(360deg); } }

.aj-menu {
  visibility: hidden;
  position: absolute;
  top: 50%;
  right: 50%;
  background-color: #ffffff;
  border-radius: 0.3rem;
  box-shadow: none;
  z-index: 999;
  margin: 0;
  padding: 0;
  transform: scale(0.5);
  transform-origin: top right;
  opacity: 0;
  transition: all 0.1s ease-in; }
  .aj-menu li {
    opacity: 0;
    transform: translateX(1rem);
    transition: all 0s ease-in;
    list-style-type: none;
    border-radius: 0.3rem; }
  .aj-menu a,
  .aj-menu button {
    border: none;
    background-color: #ffffff;
    padding: 0.6rem 2.8rem 0.6rem 2rem;
    font-family: 'Lato', 'Helvetica Nue', Helvetica, Arial sans-serif;
    font-weight: 400;
    color: #333333 !important;
    font-size: 1.4rem;
    display: flex;
    align-items: center;
    white-space: nowrap;
    text-decoration: none;
    margin: 0px;
    cursor: pointer; }
    .aj-menu a i,
    .aj-menu button i {
      margin-right: 1.6rem;
      color: #777777; }
    .aj-menu a span,
    .aj-menu button span {
      line-height: 1; }
    .aj-menu a:hover, .aj-menu a:focus,
    .aj-menu button:hover,
    .aj-menu button:focus {
      background-color: #f5f5f5; }
    .aj-menu a:focus,
    .aj-menu button:focus {
      outline: thin dotted #333; }
  .aj-menu-container {
    position: relative; }
    .aj-menu-container.is-active {
      overflow: initial; }
      .aj-menu-container.is-active .aj-menu,
      .aj-menu-container.is-active .aj-menu--select {
        visibility: visible;
        transform: scale(1);
        opacity: 1;
        box-shadow: 0 1px 3px rgba(0, 0, 0, 0.4);
        transition: opacity 0.15s ease-out, transform 0.15s ease-out; }
        .aj-menu-container.is-active .aj-menu li,
        .aj-menu-container.is-active .aj-menu--select li {
          opacity: 1;
          transform: translateX(0);
          transition: opacity 0.15s ease-out, transform 0.15s ease-out; }
  .aj-menu-compensation .aj-menu-container {
    opacity: 0; }
    .aj-menu-compensation .aj-menu-container > button {
      display: none; }
    .aj-menu-compensation .aj-menu-container .aj-menu {
      position: relative;
      transform: scale(1); }
  .aj-menu__divider {
    height: 0.1rem;
    width: 100%;
    background-color: #cccccc;
    margin: 0.8rem 0; }
  .aj-menu__has-status {
    display: flex;
    flex-direction: column;
    align-items: flex-start; }
    .aj-menu__has-status span:last-of-type {
      font-size: 1.2rem;
      color: #595959;
      margin-top: 0.2rem; }
  .aj-menu--right {
    right: auto;
    left: 50%;
    transform-origin: top left; }
    .aj-menu--right li {
      transform: translateX(-1rem); }
  .aj-menu--up .aj-menu {
    top: auto;
    bottom: 50%;
    transform-origin: bottom right; }
  .aj-menu--select {
    right: 0;
    top: auto;
    min-width: 20rem;
    visibility: hidden;
    position: absolute;
    background-color: #ffffff;
    border-radius: 0.3rem;
    box-shadow: none;
    z-index: 999;
    margin: 0;
    padding-top: 0.8rem;
    transform: scale(0.5);
    transform-origin: top right;
    opacity: 0;
    transition: all 0.1s ease-in; }
    .aj-menu--select li {
      opacity: 0;
      transform: translateX(1rem);
      transition: all 0s ease-in;
      list-style-type: none; }
    .aj-menu--select button {
      cursor: pointer; }
  @media (max-width: 399px) {
    .aj-menu {
      padding: 0.8rem 0; }
      .aj-menu button i,
      .aj-menu a i {
        display: none; }
      .aj-menu button span,
      .aj-menu a span {
        line-height: 2.4rem; } }

.aj-checkbox .aj-checkmark {
  background-color: #ffffff;
  padding: 0.9rem 2.4rem 1rem 6.4rem;
  font-family: 'Lato', 'Helvetica Nue', Helvetica, Arial sans-serif;
  font-weight: 400;
  color: #333333 !important;
  display: block;
  width: 100%;
  white-space: nowrap;
  cursor: pointer; }
  .aj-checkbox .aj-checkmark:hover {
    background: #f5f5f5; }

.aj-checkbox input[type="checkbox"] ~ span.aj-checkmark:before {
  top: 1rem;
  left: 2.8rem; }

.aj-checkbox input[type="checkbox"] ~ span.aj-checkmark:after {
  top: 1.1rem;
  left: 3.3rem; }

.aj-select__options {
  padding: 0 1.6rem 1.6rem; }
  .aj-select__options .aj-btn--gray {
    width: 100%;
    text-align: center; }
  .aj-select__options button {
    margin-top: 0.8rem; }

.aj-preset-item {
  display: flex;
  align-items: center;
  min-height: 3.2rem;
  padding: 0 1.6rem 0 0;
  transition: all 0.1s ease; }
  .aj-preset-item:hover {
    background-color: #f5f5f5; }

.aj-preset-item__name {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 24rem;
  border: none;
  background: none;
  font-family: 'Lato', 'Helvetica Nue', Helvetica, Arial sans-serif;
  font-weight: 400;
  color: #333333;
  font-size: 1.4rem;
  padding: 1rem 0.4rem 1rem 1.6rem;
  text-align: left;
  flex: 1;
  margin-right: auto; }

.aj-preset__btn-icon {
  border: none;
  background: none;
  height: 2.8rem;
  width: 2.8rem;
  padding: 0.4rem;
  border-radius: 0.3rem;
  margin-left: 0.4rem;
  transition: all 0.1s ease; }
  .aj-preset__btn-icon i {
    font-size: 2rem;
    color: #595959;
    transition: all 0.1s ease; }
  .aj-preset__btn-icon:hover {
    background-color: #ffffff; }
    .aj-preset__btn-icon:hover i {
      color: #333333; }

i.aj-preset__global {
  margin: 0 0.4rem;
  font-size: 2rem;
  color: #bbbbbb;
  cursor: default; }

.aj-rotate {
  animation: rotating 1s ease 0.3s infinite; }

.aj-rotate--2s {
  animation: rotating 2s linear infinite; }

@keyframes rotating {
  0% {
    transform: rotate(0deg); }
  100% {
    transform: rotate(360deg); } }
