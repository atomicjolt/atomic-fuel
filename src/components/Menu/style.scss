@import "../../styles/styles.scss";

.aj-menu {
  visibility: hidden;
  position: absolute;
  top: 50%;
  right: 50%;
  background-color: $white;
  border-radius: 0.3rem;
  box-shadow: none;
  z-index: 999;
  margin: 0;
  padding: 0;
  transform: scale(0.5);
  transform-origin: top right;
  opacity: 0;
  transition: all 0.1s ease-in;

  li {
    opacity: 0;
    transform: translateX(1rem);
    transition: all 0s ease-in;
    list-style-type: none;
    border-radius: 0.3rem;
  }

  a,
  button {
    border: none;
    background-color: $white;
    padding: 0.6rem 2.8rem 0.6rem 2rem;
    @include regular;
    color: $gray10 !important;
    font-size: 1.4rem;
    display: flex;
    align-items: center;
    white-space: nowrap;
    text-decoration: none;
    margin: 0px;
    cursor: pointer;

    i {
      margin-right: 1.6rem;
      color: $gray9;
    }

    span {
      line-height: 1;
    }

    &:hover,
    &:focus {
      background-color: $gray1;
    }
    &:focus {
      outline: thin dotted #333;
    }
  }

  &-container {
    position: relative;

    &.is-active {
      overflow: initial;

      .aj-menu,
      .aj-menu--select {
        visibility: visible;
        transform: scale(1);
        opacity: 1;
        box-shadow: 0 1px 3px rgba(0, 0, 0, 0.4);
        transition: opacity 0.15s ease-out, transform 0.15s ease-out;
        li {
          opacity: 1;
          transform: translateX(0);
          transition: opacity 0.15s ease-out, transform 0.15s ease-out;
        }
      }
    }
  }

  &-compensation &-container {
    opacity: 0;
    & > button {
      display: none;
    }
    .aj-menu {
      position: relative;
      transform: scale(1);
    }
  }

  &__divider {
    height: 0.1rem;
    width: 100%;
    background-color: $gray4;
    margin: 0.8rem 0;
  }

  &__has-status {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    span:last-of-type {
      font-size: 1.2rem;
      color: $alt-text;
      margin-top: 0.2rem;
    }
  }

  // Modifiers
  &--right {
    right: auto;
    left: 50%;
    transform-origin: top left;
    li {
      transform: translateX(-1rem);
    }
  }

  &--up {
    .aj-menu {
      top: auto;
      bottom: 50%;
      transform-origin: bottom right;
    }
  }

  &--select {
    right: 0;
    top: auto;
    min-width: 20rem;
    visibility: hidden;
    position: absolute;
    background-color: $white;
    border-radius: 0.3rem;
    box-shadow: none;
    z-index: 999;
    margin: 0;
    padding-top: 0.8rem;
    transform: scale(0.5);
    transform-origin: top right;
    opacity: 0;
    transition: all 0.1s ease-in;

    li {
      opacity: 0;
      transform: translateX(1rem);
      transition: all 0s ease-in;
      list-style-type: none;
    }
    button {
      cursor: pointer;
    }
  }

  @media (max-width: 399px) {
    padding: 0.8rem 0;

    button,
    a {
      i {
        display: none;
      }
      span {
        line-height: 2.4rem;
      }
    }
  }
}

.aj-checkbox {
  .aj-checkmark {
    background-color: $white;
    padding: 0.9rem 2.4rem 1rem 6.4rem;
    @include regular;
    color: $gray10 !important;
    display: block;
    width: 100%;
    white-space: nowrap;
    cursor: pointer;

    &:hover {
      background: $gray1;
    }
  }

  input[type="checkbox"] ~ span.aj-checkmark:before {
    top: 1rem;
    left: 2.8rem;
  }
  input[type="checkbox"] ~ span.aj-checkmark:after {
    top: 1.1rem;
    left: 3.3rem;
  }
}

.aj-select__options {
  padding: 0 1.6rem 1.6rem;

  .aj-btn--gray {
    width: 100%;
    text-align: center;
  }
  button {
    margin-top: 0.8rem;
  }
}

.aj-preset-item {
  display: flex;
  align-items: center;
  min-height: 3.2rem;
  padding: 0 1.6rem 0 0;
  transition: all 0.1s ease;

  &:hover {
    background-color: $gray1;
  }
}

.aj-preset-item__name {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  max-width: 24rem;
  border: none;
  background: none;
  @include regular;
  color: $text;
  font-size: 1.4rem;
  padding: 1rem 0.4rem 1rem 1.6rem;
  text-align: left;
  flex: 1;
  margin-right: auto;
}

.aj-preset__btn-icon {
  border: none;
  background: none;
  height: 2.8rem;
  width: 2.8rem;
  padding: 0.4rem;
  border-radius: 0.3rem;
  margin-left: 0.4rem;
  transition: all 0.1s ease;

  i {
    font-size: 2rem;
    color: $alt-text;
    transition: all 0.1s ease;
  }

  &:hover {
    background-color: $white;
    i {
      color: $text;
    }
  }
}

i.aj-preset__global {
  margin: 0 0.4rem;
  font-size: 2rem;
  color: $gray5;
  cursor: default;
}

.aj-rotate {
  animation: rotating 1s ease 0.3s infinite;
}

.aj-rotate--2s {
  animation: rotating 2s linear infinite;
}

@keyframes rotating {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
