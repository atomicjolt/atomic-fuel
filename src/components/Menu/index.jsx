import { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';

import { useListenOutsideClick } from '../common/hooks';
import './style.css';
export default function Menu(props) {
  const [menuOpen, setMenuOpen] = useState(false);
  const activeClass = menuOpen ? 'is-active' : '';
  const { renderChildren } = props;

  const closeMenu = () => {
    setMenuOpen(false);
  };

  const onKeyPress = (e) => {
    if ('keyCode' in e && e.keyCode === 13) {
      closeMenu();
    }
  };

  const toggleMenu = () => {
    setMenuOpen(!menuOpen);
  };

  const { ref: menuRef } = useListenOutsideClick(closeMenu);

  return renderChildren(
    activeClass,
    closeMenu,
    menuOpen,
    menuRef,
    onKeyPress,
    toggleMenu,
  );
}

Menu.propTypes = {
  renderChildren: PropTypes.func.isRequired,
};
