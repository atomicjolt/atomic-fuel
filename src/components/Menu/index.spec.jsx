import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Menu from '.';

/* eslint-disable react/prop-types */
const MockChild = (props) => {
  const {
    activeClass,
    closeMenu,
    menuRef,
    toggleMenu,
  } = props;

  return (
    <div className={activeClass}>
      <button id="buttonToggle" onClick={toggleMenu}>
        Open
      </button>
      <ul ref={menuRef}>
        <li>
          Menu Item
        </li>
        <li>
          <button id="buttonClose" onClick={closeMenu}>
            Close
          </button>
        </li>
      </ul>
    </div>
  );
};

describe('Menu component', () => {

  const renderComp = () => render(
    <Menu renderChildren={(activeClass, closeMenu, menuOpen, menuRef, toggleMenu) => (
      <MockChild
        activeClass={activeClass}
        toggleMenu={toggleMenu}
        closeMenu={closeMenu}
        menuOpen={menuOpen}
        menuRef={menuRef}
      />
    )}
    />);

  it('Should match snapshot when given renderChildren prop', () => {
    const { asFragment } = renderComp();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should open the menu', () => {
    renderComp();
    fireEvent.click(screen.getByText('Open'));
    expect(screen.queryByText('Menu Item')).toBeTruthy();
  });
});
