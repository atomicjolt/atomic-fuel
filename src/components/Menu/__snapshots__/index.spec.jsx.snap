// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component Should match snapshot when given renderChildren prop 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      id="buttonToggle"
    >
      Open
    </button>
    <ul>
      <li>
        Menu Item
      </li>
      <li>
        <button
          id="buttonClose"
        >
          Close
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
